pkg/auth/authenticator.go

```diff
@@ -69,3 +71,46 @@ func (ba *BuiltinAuthenticator) AuthenticateUser(ctx context.Context, username,
 func (ba *BuiltinAuthenticator) String() string {
 	return "built in authenticator"
 }
+
+// GetOrCreateUser searches for the user by username, and if not found, creates a new user with the given username and
+// external user identifier.  It returns the username of the user.
+// This function is meant to be used by Authenticator implementations in the ChainAuthenticator.
+func GetOrCreateUser(ctx context.Context, authService Service, username, identifier, defaultUserGroup, source string) (string, error) {
```

Comment:
This function is a bit confusing as a generic one. 
- The identifier -> should be simple, the username, the caller should decide what the username is. 
- The username is only used for friendly name, so just be explicit about it at the function signature.
- No need Don't be opinionated about returning string
```suggestion
func GetOrCreateUser(ctx context.Context, authService Service, username, friendlyName, defaultUserGroup, source string) (*model.User, error) {
```

Reply:
> The identifier -> should be simple, the username, the caller should decide what the username is.

Didn't understand this comment, but the rest is done

Reply:
lol english very very hard (explain in further comment)